### PR TITLE
[Ldap] Ldap Entry case-sensitive attribute key option

### PR DESCRIPTION
**Pull request by karlshea**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

See PR #36432

### CODE REVIEW

File: src/Symfony/Component/Ldap/Entry.php [hunk 1]
```diff
@@ -13,16 +13,22 @@
...
 class Entry
 {
     private $dn;
     private $attributes;
+    private $lowerMap;
 
     public function __construct(string $dn, array $attributes = [])
     {
         $this->dn = $dn;
-        $this->attributes = $attributes;
```
Comment (jderusse) [inline src/Symfony/Component/Ldap/Entry.php:25]: should be initialized to an empty array

**Review by jderusse — CHANGES_REQUESTED** [review]
Thank you for this PR.

This is not a bug fix, but a new feature, and should target Symfony 5.x

**Comment by karlshea** [comment]
@jderusse Thanks!

The reason for this PR is for the Drupal LDAP module, and the issue is here: https://www.drupal.org/project/ldap/issues/3126807

That module is using 4.4. Are you saying that this PR could only appear in 5.x? Or that I should target 5.x and it could be backported?

**Comment by derrabus** [comment]
> Are you saying that this PR could only appear in 5.x? 

Yes. 4.4 is closed for features, so you would need to target 5.x.

**Comment by karlshea** [comment]
Just to be clear so I can update the Drupal issue, there is no possibility of this appearing in 4.x?

File: src/Symfony/Component/Ldap/Entry.php [hunk 2]
```diff
@@ -38,13 +45,20 @@ public function getDn()
...
      *
-     * @param string $name The name of the attribute
+     * @param string $name          The name of the attribute
+     * @param bool   $caseSensitive Whether the check should be case-sensitive
      *
      * @return bool
      */
-    public function hasAttribute($name)
+    public function hasAttribute($name, $caseSensitive = true)
     {
```
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]:
```suggestion
    public function hasAttribute($name /* , $caseSensitive = true */)
    {
        $caseSensitive = 2 > \func_num_args() || true === \func_get_arg(1);
```

Adding a parameter to the signature of a public method is a BC break. You can document and use it already, but we have to wait until Symfony 6 to add it to the signature.

File: src/Symfony/Component/Ldap/Entry.php [hunk 3]
```diff
@@ -53,13 +67,20 @@ public function hasAttribute($name)
...
      * this value is returned as an array.
      *
-     * @param string $name The name of the attribute
+     * @param string $name          The name of the attribute
+     * @param bool   $caseSensitive Whether the attribute name is case-sensitive
      *
      * @return array|null
      */
-    public function getAttribute($name)
+    public function getAttribute($name, $caseSensitive = true)
```
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]: Same problem.

File: src/Symfony/Component/Ldap/Entry.php [hunk 4]
```diff
@@ -90,5 +112,23 @@ public function setAttribute($name, array $value)
...
+
+    /**
+     * Get the attribute key.
+     *
+     * @param string $name          The attribute name
+     * @param bool   $caseSensitive Whether the attribute name is case-sensitive
+     *
+     * @return string|null
+     */
+    private function getAttributeKey($name, $caseSensitive = true)
```
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]:
```suggestion
     */
    private function getAttributeKey(string $name, bool $caseSensitive = true): ?string
```

File: src/Symfony/Component/Ldap/Entry.php [hunk 5]
```diff
@@ -90,5 +112,23 @@ public function setAttribute($name, array $value)
...
+    {
+        if ($caseSensitive) {
+            return $name;
+        }
+
+        return isset($this->lowerMap[strtolower($name)]) ? $this->lowerMap[strtolower($name)] : null;
```
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]:
```suggestion
        return $this->lowerMap[strtolower($name)] ?? null;
```

File: src/Symfony/Component/Ldap/Entry.php [hunk 6]
```diff
@@ -53,13 +67,20 @@ public function hasAttribute($name)
...
      *
      * @return array|null
      */
-    public function getAttribute($name)
+    public function getAttribute($name, $caseSensitive = true)
     {
-        return isset($this->attributes[$name]) ? $this->attributes[$name] : null;
+        $attributeKey = $this->getAttributeKey($name, $caseSensitive);
+
+        if (!$attributeKey) {
```
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]:
```suggestion
        if (null === $attributeKey) {
```

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Ldap/Entry.php (hunk 2 above)
Comment (karlshea) [reply src/Symfony/Component/Ldap/Entry.php]: Is the Symfony code style to comment out the parameter like you have in the suggestion? Or should I remove it entirely?

File: src/Symfony/Component/Ldap/Entry.php [hunk 7]
```diff
@@ -38,13 +45,21 @@ public function getDn()
     /**
      * Returns whether an attribute exists.
      *
-     * @param string $name The name of the attribute
+     * @param string $name          The name of the attribute
+     * @param bool   $caseSensitive Whether the check should be case-sensitive
      *
      * @return bool
      */
     public function hasAttribute($name)
```
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]: Please the new parameter as a comment already, as suggested in my previous review. This way, we don't forget to add it in 6.0.

File: src/Symfony/Component/Ldap/Entry.php (hunk 2 above)
Comment (derrabus) [reply src/Symfony/Component/Ldap/Entry.php]: Yes, the comment reminds us to enable the parameter on the next major version.

**Comment by derrabus** [comment]
> Just to be clear so I can update the Drupal issue, there is no possibility of this appearing in 4.x?

There is no 4.x anymore. We're only bugfixing 4.4 and feature development has shifted to 5.x. I cannot make a final decision here, but to me this change does not look like a bugfix. Is it a big problem to support 5.x as well?

**Comment by karlshea** [comment]
> There is no 4.x anymore. We're only bugfixing 4.4 and feature development has shifted to 5.x. I cannot make a final decision here, but to me this change does not look like a bugfix. Is it a big problem to support 5.x as well?

I don't think Drupal 8/Drupal 9 support Symfony 5.x, so that module is stuck with 4.4. They're already using a fork that only contains the Ldap components though, so I'm not sure what direction the maintainer will eventually go.

**Comment by derrabus** [comment]
> I don't think Drupal 8/Drupal 9 support Symfony 5.x, so that module is stuck with 4.4.

The LDAP component is pretty much self-contained. You should be able to upgrade `symfony/ldap` to 5.2 while pinning all other components to 4.4.

**Comment by karlshea** [comment]
Good to know! I'll follow up in the Drupal issue and see what they want to do.

Current test failure seems unrelated to Ldap.

**Review by derrabus — APPROVED** [review]
(for 5.x)

**Review by jderusse — APPROVED** [review]

**Comment by derrabus** [comment]
@karlshea I've force-pushed to your branch in order to resolve conflicts with 5.x and to test your changes with our new LDAP integration test suite.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @karlshea.

**Pull request merged**